src/FileSystem/SourceFileFilter.php

```diff
@@ -0,0 +1,119 @@
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\FileSystem;
+
+use function explode;
+use Infection\FileSystem\Finder\Iterator\RealPathFilterIterator;
+use Infection\TestFramework\Coverage\CoveredFileData;
+use Iterator;
+use function Pipeline\take;
+use Symfony\Component\Finder\SplFileInfo;
+
+/**
+ * @internal
+ */
+final class SourceFileFilter
+{
+    private $sourceDirectories;
+    private $excludeDirectories;
+    /** @var string[] */
+    private $filters;
+
+    /**
+     * @param string[] $sourceDirectories
+     * @param string[] $excludeDirectories
+     *
+     * @return iterable<SplFileInfo>
+     */
+    public function __construct(
+        array $sourceDirectories,
+        array $excludeDirectories,
+        string $filter
+    ) {
+        $this->sourceDirectories = $sourceDirectories;
+        $this->excludeDirectories = $excludeDirectories;
+        $this->filters = take(explode(',', $filter))
+            ->map('trim')
+            ->filter()
+            ->toArray();
+    }
+
+    /**
+     * @return string[]
+     *
+     * @see SourceFileCollector
+     */
+    public function getFilters(): array
+    {
+        return $this->filters;
+    }
+
+    /**
+     * @param iterable<SplFileInfo&CoveredFileData> $input
+     *
+     * @return iterable<SplFileInfo&CoveredFileData>
+     */
+    public function filter(iterable $input): iterable
+    {
+        if ($this->filters === []) {
+            return $input;
+        }
+
+        return new RealPathFilterIterator(
+            $this->iterableToIterator($input),
+            $this->filters,
+            []
+        );
+    }
+
+    /**
+     * @param iterable<SplFileInfo&CoveredFileData> $input
+     *
+     * @return \Iterator<SplFileInfo&CoveredFileData>
```

Comment:
```suggestion
     * @return Iterator<SplFileInfo&CoveredFileData>
```
(it has already been imported)